Add and use email-sending domain

dav3r opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##
This PR adds:
* A new Terraform input variable for the email-sending domain to be used in assessments
* A cloud-init script that sets the `PRIMARY_DOMAIN` variable for the postfix service in the PCA-Gophish Docker composition to the value of the Terraform email-sending domain variable

<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##
As specified in #110:
> Many of our assessments have the need to send emails from a specific domain, which currently requires several manual configuration steps. This would automate part of that process.

Resolves #110.
<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

I applied these Terraform changes in a COOL environment and confirmed that:
* The Postfix container's primary domain was correctly set based on the value of the Terraform input variable for the email-sending domain.
* The Postfix container and the pca-gophish-composition SystemD service on the Gophish instance started up correctly.
* Email was successfully sent via Postfix in the assessment environment from the email-sending domain.

<!-- How did you test your changes? How could someone else test this PR? -->

<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

## ✅ Checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

* [x] This PR has an informative and human-readable title.
* [x] Changes are limited to a single goal - _eschew scope creep!_
* [x] All relevant type-of-change labels have been added.
* [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
* [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
* [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
* [x] All new and existing tests pass.

dav3r committed:
Add new variable for email-sending domain

dav3r committed:
Create a cloud-init script template to set the email-sending domain

The script modifies the PRIMARY_DOMAIN postfix parameter in the 
pca-gophish Docker composition file.

dav3r committed:
Add the postfix-setup script to the gophish cloud-init user data

jsf9k commented on README.md:
```diff
@@ -95,6 +95,7 @@ the COOL environment.
 | cert_bucket_name | The name of the AWS S3 bucket where certificates are stored. | `string` | `cisa-cool-certificates` | no |
 | cool_domain | The domain where the COOL resources reside (e.g. "cool.cyber.dhs.gov"). | `string` | `cool.cyber.dhs.gov` | no |
 | dns_ttl | The TTL value to use for Route53 DNS records (e.g. 86400).  A smaller value may be useful when the DNS records are changing often, for example when testing. | `number` | `60` | no |
+| email_sending_domain | The domain to send emails from within the assessment environment (e.g. "example.com"). | `string` | `example.com` | no |
```
Do you mean the domain or the hostname?  If the user wants to send email from `treeofwoe.thulsa-doom.com` is that the value they specify for this variable?  Or do they just specify the domain `thulsa-doom.com` and the `treeofwoe` is specified elsewhere?

jsf9k replied:
If the user is supposed to specify the hostname then I suggest changing the name of the variable to make this clearer.  At a minimum, you should change the default to `mail.example.com` if the user is expected to specify the hostname instead of just the domain.

dav3r replied:
Honestly, I don't know how much this value really matters.  If you set it to `thulsa-doom.com`, you can successfully send mail from either `conan@thulsa-doom.com` or `conan@treeofwoe.thulsa-doom.com`.  I did a test of this and it looks like as long as there is a valid MX record for `thulsa-doom.com` and `treeofwoe.thulsa-doom.com`, either one should be acceptable.

That is why I worded it as `domain` instead of `hostname` or `fully-qualified domain name`.

PS: I used https://mail-tester.com to get a score (with explanations) for the test emails I sent and they were scored the same, regardless of whether they were sent from `conan@thulsa-doom.com` or `conan@treeofwoe.thulsa-doom.com`.  For these tests, my `email_sending_domain` was just the `thulsa-doom.com` domain (not really, but the domain that I have been using for my testing).

jsf9k replied:
Perfect.  I withdraw the question.

hillaryj reviewed (approved):
🗡️

jsf9k commented on cloud-init/postfix-setup.tpl.yml:
```diff
@@ -0,0 +1,6 @@
+runcmd:
+  # Modify postfix's PRIMARY_DOMAIN in pca-gophish-composition docker-compose
+  # file to the email-sending domain for this assessment
+  - [sed, -i, "s|PRIMARY_DOMAIN=.*|PRIMARY_DOMAIN=${email_sending_domain}|g",
```
This regex will apply to lines that look like `.*PRIMARY_DOMAIN=.*$`.  I suspect that you want it to apply only to lines that look like `^PRIMARY_DOMAIN=[^\w]*$`.  I suggest changing the regex to this to be extra safe:
```suggestion
  - [sed, -i, "s|^PRIMARY_DOMAIN=[^\w]*$|PRIMARY_DOMAIN=${email_sending_domain}|g",
```

mcdonnnj replied:
Based on the `docker-compose.yml` this is likely editing, at least part of your edit would not work: https://github.com/cisagov/pca-gophish-composition/blob/47c69034a5df627852444932ed960660aeadce2b/docker-compose.yml#L67

jsf9k replied:
Ah, your eyes are as sharp as ever @mcdonnnj!  How about this then:
```suggestion
  - [sed, -i, "s|^\([- \w]*\)PRIMARY_DOMAIN=[^\w]*$|\1PRIMARY_DOMAIN=${email_sending_domain}|g",
```

dav3r replied:
Thanks @mcdonnnj, you beat me to the punch on the `docker-compose.yml` stuff. 😄  

@jsf9k, your most-recent suggestion is almost perfect, but I had to escape the backslashes to appease the Python YAML linter used by cloud-init as well as add a `\` linebreak to satisfy our pre-commit `yamllint` hook.  

I verified that it worked as expected by deploying it to my test environment, then made the change in https://github.com/cisagov/cool-assessment-terraform/pull/112/commits/4d2c4c29757ea4fabfd570a78323e2f40c78408b.

Thanks for the improvement!

jsf9k reviewed (changes requested):
I have a couple of questions.

dav3r committed:
Make regex more specific

This reduces the chances of any unintentional matches.  Thanks @jsf9k 
for the improvement suggestion!

Co-Authored-By: Shane Frasier <jeremy.frasier@trio.dhs.gov>

jsf9k reviewed (approved):
Great work!  Thanks for performing a real test of the suggested regex changes.  I tested locally by `echo`ing various strings and piping that into `sed`, but a real test is vastly better.

dav3r commented:
> Great work! Thanks for performing a real test of the suggested regex changes. I tested locally by `echo`ing various strings and piping that into `sed`, but a real test is vastly better.

FYI, after a couple of rounds of trying to fix the YAML errors (requiring a redeploy of my Gophish instance each time), I discovered [this handy bit of Python from cloud-init](https://github.com/canonical/cloud-init/blob/master/tools/validate-yaml.py), which allowed me to quickly verify locally that all was well.  After I got that `validate-yaml.py` script to confirm that my YAML was valid, I redeployed my instance with it and confirmed that everything was groovy.
